Autopilot/GPS/MTK33x9.c

```diff
@@ -0,0 +1,385 @@
+            uart_buffer[buffer_index] = data;
+            buffer_index = (buffer_index + 1) % GPS_UART_BUFFER_SIZE; //make sure we dont cause a memory fault here
+        }
+    }
+    IFS0bits.U1RXIF = 0; // Clear the Interrupt Flag
+}
+
+/**
+ * Given an NMEA string starting after the $, verifies the integrity of the stirng
+ * using the checksum
+ * @param string
+ * @return True if string is a valid gps string, false otherwise
+ */
+static bool isNMEAChecksumValid(char* string){
+    uint16_t i = 0;
+    uint8_t checksum = 0;
+    
+    while(string[i] != '*'){
+        checksum ^= string[i];
+        i++;
+    }
+    i++;
+
+    return byteToHexString((checksum & 0xF0) >> 4) == string[i] &&  byteToHexString(checksum & 0x0F) == string[i+1];
+}
+
+static void parseVTG(char* data){
+    //static so that we dont allocate these variables every time
+    static char rawHeading[6] = {0, 0, 0, 0, 0, 0};
+    static char rawGroundSpeed[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+    
+    int comma = 0;
+    int i = 0;
+    int j = 6;
+
+    while (data[j] != '*') {
+        char numData = asciiToHex(data[j]);
+        if (data[j] == ',') {
+            comma++;
+            i = 0;
+        }
+        if (comma == 1 && (i != 0)) {
+            rawHeading[i] = numData;
+        }
+        if (comma == 7 && (i != 0)) {
+            rawGroundSpeed[i] = numData;
+        }
+        i++;
+        j++;
+    }
+
+    i = 1;
+    long int multiplier = 10;
+    int decimalPoint = 0;
+    
+    gps_data.heading = 0;
+    float tHeading = 0;
+    for (i = 1; i < 6; i++) //this code first generates an 5 digit decimal number
+    {
+        if (rawHeading[i] == 0x10)//check for decimal point
+        {
+            decimalPoint = i;
+        } else {
+            tHeading += (float) (rawHeading[i]*100000 / multiplier);
+            multiplier *= 10;
+        }
+    }
+    decimalPoint = decimalPoint - 2;
+    multiplier = 10000;
+    while (decimalPoint > 0) //then divdes it according to the placement of the decimal
+    {
+        multiplier = multiplier / 10;
+        decimalPoint--;
+    }
+    gps_data.heading = (int)(tHeading / multiplier);
+
+    //	//calculate speed - tricky because of unknown 1-3 digits preceeding the decimal
+    i = 1;
+    multiplier = 10;
+    decimalPoint = 0;
+    gps_data.ground_speed = 0;
+    for (i = 1; i < 7; i++) //this code first generates an 6 digit decimal number
+    {
+        if (rawGroundSpeed[i] == 0x10)//check for decimal point
+        {
+            decimalPoint = i;
+        } else {
+            gps_data.ground_speed += (float) (rawGroundSpeed[i]*1000000 / multiplier);
+            multiplier = multiplier * 10;
+        }
+    }
+    decimalPoint = decimalPoint - 2;
+    multiplier = 100000;
+    while (decimalPoint > 0) //then divdes it according to the placement of the decimal
+    {
+        multiplier = multiplier / 10;
+        decimalPoint--;
+    }
+    gps_data.ground_speed = gps_data.ground_speed / multiplier;
```

Comment:
Yes the GPS module itself does, we do the conversion (i guess to avoid calculations?)